### PR TITLE
CUDA device API & VerifyGPUCode pass update

### PR DESCRIPTION
**Pull request by jcf94**
This pr is part of #5883 , consists of 2 small improvements.

1. Add `kMaxRegistersPerBlock` to device api, mainly used for CUDA api to get the local memory limitation;

2. Add vectorize check to VerifyGPUCode pass depending on the vectorized block data size, nvcc has more strict limitations than llvm

The rest 2 bug fix will be in seperate PRs.
~~~
3. Update RewriteSimplify to fix a bug occurred in vectorized GPU shared memory cooperative fetching.
Test cases are in `tests/python/unittest/test_te_schedule_gpu_advanced.py`, see `test_vectorized_cooperative_fetching_x()` and `test_vectorized_cooperative_fetching_xy()` for more information.

Code generated with bug is shown like this:
```
A.shared[ramp(((ax0.ax1.fused.outer.outer*256) + (threadIdx.x_1*4)), 1, 4)] =
(float32x4*)A_2[(((broadcast(((floordiv(blockIdx.x, 4)*32768) + (ax0.ax1.fused.outer.outer*2048)), 4) + (floordiv(ramp((threadIdx.x_1*4), 1, 4), broadcast(64, 4))*broadcast(512, 4))) + broadcast((k.outer.outer*64), 4)) + floormod(ramp((threadIdx.x_1*4), 1, 4), broadcast(64, 4)))])
```
Which will finally lower to wrong CUDA C instructions.
This should be simplified to generate the correct RampNode:
```
A.shared[ramp(((ax0.ax1.fused.outer.outer*256) + (threadIdx.x_1*4)), 1, 4)] =
(float32x4*)A_2[ramp((((((floordiv(blockIdx.x, 4)*32768) + (ax0.ax1.fused.outer.outer*2048)) + (floordiv(threadIdx.x_1, 16)*512)) + (k.outer.outer*64)) + (floormod(threadIdx.x_1, 16)*4)), 1, 4)])
```
4. Add LegalizeInvalidAttach in `schedule_dataflow_rewrite.cc`
This function legalizes the compute_at location if the target iterator of compute_at was split or fused. The following two cases will crash if we don't use this function.

```python
import tvm
from tvm import te

A = te.compute((10, 10), lambda i, j: 1.0, name='A')
B = te.compute((10, 10), lambda i, j: A[i][j], name='B')

# Case 1: Split an axis which is the target of a compute_at
s = te.create_schedule([B.op])
s[A].compute_at(s[B], B.op.axis[1])
s[B].split(B.op.axis[1], 2)

print(tvm.lower(s, [A, B], simple_mode=True))

# Case 2: Fuse an axis which is the target of a compute_at
s = te.create_schedule([B.op])
s[A].compute_at(s[B], B.op.axis[1])
s[B].fuse(B.op.axis[0], B.op.axis[1])

print(tvm.lower(s, [A, B], simple_mode=True))
```

We rebuild the attach relation in`Schedule::normalize()` to fix this.
~~~

### CODE REVIEW

**Comment by jcf94** [comment]
Seems the extra simplify rule caused a UT error, I'll try to fix that.

File: src/te/schedule/schedule_dataflow_rewrite.cc [hunk 1]
```diff
@@ -614,10 +614,74 @@ void InjectInline(ScheduleNode* sch) {
   }
 }
 
+void LegalizeInvalidAttach(ScheduleNode* sch) {
```
Comment (merrymercy) [inline src/te/schedule/schedule_dataflow_rewrite.cc:617]:
```suggestion
void LegalizeInvalidAttach(ScheduleNode* sch) {
  // This legalizes the compute_at location if the target iterator of compute_at was split or fused.
  // Case 1: If the target of compute_at is split, we will move the compute_at location to the inner iterator.
  // Case 2: If the target of compute_at is fused, we will move the compute_at location to the newly fused iterator.
  // Note that case 2 can only happen if the target of compute_at is the innermost iterator of fuse operation.
```

Document this function

**Comment by merrymercy** [comment]
There are very simple unit tests for `LegalizeInvalidAttach`

```python
import tvm
from tvm import te

A = te.compute((10, 10), lambda i, j: 1.0, name='A')
B = te.compute((10, 10), lambda i, j: A[i][j], name='B')

# Case 1: Split an axis which is the target of a compute_at
s = te.create_schedule([B.op])
s[A].compute_at(s[B], B.op.axis[1])
s[B].split(B.op.axis[1], 2)

print(tvm.lower(s, [A, B], simple_mode=True))

# Case 2: Fuse an axis which is the target of a compute_at
s = te.create_schedule([B.op])
s[A].compute_at(s[B], B.op.axis[1])
s[B].fuse(B.op.axis[0], B.op.axis[1])

print(tvm.lower(s, [A, B], simple_mode=True))
```

File: src/arith/rewrite_simplify.cc [hunk 2]
```diff
@@ -865,6 +900,28 @@ PrimExpr RewriteSimplifier::Impl::VisitExpr_(const FloorModNode* op) {
     TVM_TRY_REWRITE_IF(floormod(x + y * c1, c2), floormod(x, c2),
                        c2.Eval()->value > 0 && c1.Eval()->value % c2.Eval()->value == 0);
 
+    TVM_TRY_REWRITE_IF(floormod(x * c1 + y, c2), floormod(x, floordiv(c2, c1)) * c1 + y,
+                       c1.Eval()->value > 0 && c2.Eval()->value > 0 &&
+                           c2.Eval()->value % c1.Eval()->value == 0 &&
+                           CanProveGreaterEqual(-y.Eval(), -c1.Eval()->value + 1));
+
+    // TODO(jcf94): For the next two rules, better use the max common factor
+    // of c1, c2, c3 to do the simplification
```
[the remark this answers is not in the document]
Comment (jcf94) [reply src/arith/rewrite_simplify.cc:909]: Emm ... Unfortunately these rules seemed to cause another simplify error int the UT(see jenkins tests), I'll try to see if we can figure out a better way rule to solve it fine.

File: tests/python/unittest/test_te_schedule_gpu_advanced.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    s[AA].bind(aa1, tvm.te.thread_axis("threadIdx.y"))
+    s[AA].bind(aa2, tvm.te.thread_axis("threadIdx.x"))
+    s[AA].vectorize(aa3)
+    s[BB].bind(bb1, tvm.te.thread_axis("threadIdx.y"))
+    s[BB].bind(bb2, tvm.te.thread_axis("threadIdx.x"))
+    s[BB].vectorize(bb3)
+
+    vcf_check_common(s, [A, B, C])
+
+def test_cache_write_follow_split_fuse():
```
Comment (merrymercy) [inline tests/python/unittest/test_te_schedule_gpu_advanced.py:147]: We can use the simpler version of unit test for `LegalizeInvalidAttach` to replace this.

**Comment by tqchen** [comment]
Please split the rewrite simplify into its own PR and add unittestcases for the newly added rules. It would be great if we can write down a few lines of proof sketches, to make sure that these rules are correct.

**Comment by jcf94** [comment]
> Please split the rewrite simplify into its own PR and add unittestcases for the newly added rules. It would be great if we can write down a few lines of proof sketches, to make sure that these rules are correct.

Thanks, I'll reconsider about the simplify rules to make sure it wouldn't bring extra bugs and split it to another PR.

File: src/tir/analysis/verify_gpu_code.cc [hunk 4]
```diff
@@ -62,6 +64,9 @@ class GPUCodeVerifier : public StmtVisitor {
       size_t size = static_cast<size_t>(op->constant_allocation_size());
       shared_memory_per_block_ += size * op->dtype.bytes() * op->dtype.lanes();
     }
+    if (op->dtype.lanes() > 1) {
+      valid_ &= op->dtype.lanes() * op->dtype.bytes() <= static_cast<int>(max_vector_bytes_);
```
Comment (merrymercy) [inline src/tir/analysis/verify_gpu_code.cc:68]: Cast INT64_MAX to int will overflow. You should cast the LHS.

File: src/tir/analysis/verify_gpu_code.cc [hunk 5]
```diff
@@ -129,6 +134,17 @@ class GPUCodeVerifier : public StmtVisitor {
     }
   }
 
+  void VisitExpr_(const LoadNode* op) {
+    // Currently not able to check out: If the index expression failed
+    // to be simplified to a RampNode
+    if (op->index->IsInstance<RampNode>()) {
+      if (op->dtype.lanes() > 1) {
+        valid_ &= op->dtype.lanes() * op->dtype.bytes() <= static_cast<int>(max_vector_bytes_);
```
Comment (merrymercy) [inline src/tir/analysis/verify_gpu_code.cc:142]: Same as the above.

**Comment by jcf94** [comment]
@merrymercy cast fix & CI clean

**Pull request merged**